src/coreclr/scripts/superpmi.py

```diff
@@ -625,67 +647,67 @@ def __merge_mch_files__(self):
             proc = subprocess.Popen(command)
-
             proc.communicate()
 
-        if not os.path.isfile(self.mcs_path):
-            raise RuntimeError("mch file failed to be generated at: %s" % self.mcs_path)
+        if not os.path.isfile(self.base_mch_file):
+            raise RuntimeError("MCH file failed to be generated at: %s" % self.base_mch_file)
+
+    def __create_thin_unique_mch__(self):
+        """  Create a thin unique MCH
+        
+        Notes:
+            <mcl> -removeDup -thin <s_baseMchFile> <s_nodupMchFile>
+        """
+
+        command = [self.mcs_path, "-removeDup", "-thin", self.base_mch_file, self.nodup_mch_file]
+        print("Invoking: " + " ".join(command))
+        proc = subprocess.Popen(command)
+        proc.communicate()
+
+        if not os.path.isfile(self.nodup_mch_file):
+            raise RuntimeError("Error, no dup mch file not created correctly at: %s" % self.nodup_mch_file)
+
+        if not self.coreclr_args.skip_cleanup:
+            os.remove(self.base_mch_file)
+            self.base_mch_file = None
     
     def __create_clean_mch_file__(self):
-        """ Create a clean mch file based on the original
+        """ Create a clean mch file
 
         Notes:
-            <SuperPMIPath> -p -f <s_baseFailMclFile> <s_baseMchFile> <jitPath>
+            <SuperPMIPath> -p -f <s_baseFailMclFile> <s_nodupMchFile> <jitPath>
 
             if <s_baseFailMclFile> is non-empty:
-                <mcl> -strip <s_baseFailMclFile> <s_baseMchFile> <s_cleanMchFile>
+                <mcl> -strip <s_baseFailMclFile> <s_nodupMchFile> <s_finalMchFile>
             else
-                # no need to copy, just change the names
-                clean_mch_file = base_mch_file
+                # copy/move nodup file to final file
             del <s_baseFailMclFile>
         """
 
-        command = [self.superpmi_path, "-p", "-f", self.base_fail_mcl_file, self.base_mch_file, self.jit_path]
-        print (" ".join(command))
+        command = [self.superpmi_path, "-p", "-f", self.base_fail_mcl_file, self.nodup_mch_file, self.jit_path]
+        print("Invoking: " + " ".join(command))
         proc = subprocess.Popen(command)
-
         proc.communicate()
 
-        if os.path.isfile(self.base_fail_mcl_file) and os.stat(self.base_fail_mcl_file).st_size != 0:
-            command = [self.mcs_path, "-strip", self.base_fail_mcl_file, self.base_mch_file, self.clean_mch_file]
-            print (" ".join(command))
+        if is_nonzero_length_file(self.base_fail_mcl_file):
+            command = [self.mcs_path, "-strip", self.base_fail_mcl_file, self.nodup_mch_file, self.final_mch_file]
+            print("Invoking: " + " ".join(command))
             proc = subprocess.Popen(command)
-
             proc.communicate()
         else:
-            self.clean_mch_file = self.base_mch_file
-            self.base_mch_file = None
+            # Ideally we could just rename this file instead of copying it.
```

Comment:
why is not `final_mch_file = nodup_mch_file`  suitable now?

Reply:
Because the "final file" is a file name or path specified by the user, so we need to copy the nodup file to that specified location. When it was an intermediate step, we could just change the variables because we would eventually generate the final file.